### ISSUE TITLE
Closure inside getStateUsing() called 3 times per row

### ISSUE THREAD

**Issue by jonaaix** [laravel-filament/filament#14852]
### Package

filament/tables

### Package Version

v3.2.124

### Laravel Version

v11.33.2

### Livewire Version

default

### PHP Version

8.3.13

### Problem description

When using `->getStateUsing()` to fill a column with a custom value, the Closure seems to be called 3 times per row!
<img width="825" alt="image" src="https://github.com/user-attachments/assets/3007e23d-ba5d-4c07-a331-563f2c1c8d2f">

**If this closure contains a custom query, this will lead to triple query execution per row!**
<img width="1263" alt="Bildschirmfoto 2024-11-20 um 16 16 22" src="https://github.com/user-attachments/assets/bf33e0f6-b75c-4fe8-bc1a-7d73d03b0832">

### Expected behavior

The Closure should be executed once on page load

### Steps to reproduce

Example Column:

```php
Tables\Columns\TextColumn::make('permission_number')
  ->label('Permissions')
  ->badge()
  ->getStateUsing(function (User $user) {
      return $user->permissions()->count();
  }),
```

### Reproduction repository (issue will be closed if this is not valid)

[https://github.com/btxtiger/filament-getstateusing-bug-14852](https://github.com/btxtiger/filament-getstateusing-bug-14852)

**Comment by Cybrarist** [comment]
i think you should use formatStateUsing instead of getStateUsing()

**Comment by jonaaix** [comment]
@Cybrarist this does not work, since the state of the column should be _**any**_ query or value (probably also a result of an HTTP request, Cache, other service, ...), and not dependant on a model column.

**Comment by Cybrarist** [comment]
@btxtiger  oh ok, i get what you mean, then why not doing it like the following 

```php 
$table->modifyQueryUsing(function (Builder $query) {
      $query->withCount('permissions');
})
```

and then you use 'permissions_count' in your TextColumn.

**Comment by jonaaix** [comment]
@Cybrarist yes this would be a workaround for this specific example, but would not solve the actual problem.
Let me show you a different example:

```php
Tables\Columns\TextColumn::make('permission_granted')
  ->label('Permissions')
  ->badge()
  ->getStateUsing(function (User $user) {
      return Http::get('https://dummy-auth-provider.com/api/check-permissions', ['uuid' => $user->uuid])
          ->json('permissions_granted') ? 'Granted' : 'Not Granted';
  });
```

In this case, the closure is still executed 3 times, and thus make 3 https requests * number of rows, for 1 page load

**Comment by ericmp33** [comment]
https://github.com/filamentphp/filament/issues/14852#issuecomment-2490981394

Seems a very bad idea to get the state of a column by doing an HTTP call for each one.

**Comment by jonaaix** [comment]
> Seems a very bad idea to get the state of a column by doing an HTTP call for each one.

Yeah, but still not targeting the issue. The issue is 3 redundant calls of a function where only one should be. Filament 3 became so much slower (less snappy) than v2, and I suspect this is one issue

**Comment by bilalelhaj** [comment]
have the same problem here

**Comment by mc-fu** [comment]
can confirm this (using function to calculate some values) - is run 3 times and output triples the value...

**Comment by borjajimnz** [comment]
After reviewing the code, it seems the issue might be associated with the `copyable` functionality. While I’m not deeply familiar with FilamentPHP, I can suggest that disabling the click should prevent the replication issue:  

```php
->disabledClick(true)
```  

I hope this helps as a temporary workaround until a proper solution is found. thanks.

**Comment by robinmoisson** [comment]
I don't know if that would have negative other consequences - any reason we don't cache the state after computing it once?

Changing [`filament/support/src/Concerns/HasCellState.php`](https://github.com/filamentphp/filament/blob/3.x/packages/support/src/Concerns/HasCellState.php#L72-L94) to 

```php
protected $cachedState = null;

public function getState(): mixed
    {
        if (! $this->getRecord()) {
            return null;
        }

        if ($this->cachedState) {
            return $this->cachedState;
        }

        $state = ($this->getStateUsing !== null) ?
            $this->evaluate($this->getStateUsing) :
            $this->getStateFromRecord();

        if (is_string($state) && ($separator = $this->getSeparator())) {
            $state = explode($separator, $state);
            $state = (count($state) === 1 && blank($state[0])) ?
                [] :
                $state;
        }

        if (blank($state)) {
            $state = $this->getDefaultState();
        }

        $this->cachedState = $state;

        return $state;
    }
```

fixes the issue for me and makes my interface really snappy again.

**Comment by robinmoisson** [comment]
Would caching the state work? Something like modifying [`packages/support/src/Concerns/HasCellState.php`](https://github.com/filamentphp/filament/blob/3.x/packages/support/src/Concerns/HasCellState.php#L72-L94) to cache the value seems to fix the issue and work well for me.

```php
protected array $cachedStatePerRecordId = [];

protected function getRecordId(): mixed
{
    return $this->getRecord()?->getKey();
}

public function getState(): mixed
{
    if (! $this->getRecord()) {
        return null;
    }

    if (array_key_exists($this->getRecordId(), $this->cachedStatePerRecordId)) {
        return $this->cachedStatePerRecordId[$this->getRecordId()];
    }

    $state = ($this->getStateUsing !== null) ?
        $this->evaluate($this->getStateUsing) :
        $this->getStateFromRecord();

    if (is_string($state) && ($separator = $this->getSeparator())) {
        $state = explode($separator, $state);
        $state = (count($state) === 1 && blank($state[0])) ?
            [] :
            $state;
    }

    if (blank($state)) {
        $state = $this->getDefaultState();
    }

    $this->cachedStatePerRecordId[$this->getRecordId()] = $state;

    return $state;
}
```

**Comment by jonaaix** [comment]
@robinmoisson Nice approach, however it sounds for me like a workaround, since it is still unclear why the execution happens three times and if it still might affect other unexposed scenarios that throttles the overall performance.

**Comment by robinmoisson** [comment]
Thanks for the feedback @btxtiger. So I dove deeper into this, and it looks like the problem is this `isDisabledClick` method which calls `getState()` to check whether the click should be disabled on this cell: https://github.com/filamentphp/filament/blob/43f74dbcc032a2b10e8365f8e4588c7224408f82/packages/tables/src/Columns/Concerns/CanBeCopied.php#L12-L22

It is called in the blade template when building the column before building the cell, I don't see an easy way to inject the state over here: https://github.com/filamentphp/filament/blob/af36dec35113be5747e12f013dd288d7287ce533/packages/tables/resources/views/index.blade.php#L1145-L1153

The state is also computed in the `text-column.blade.php` template, which I assume is the main place where it should be called: https://github.com/filamentphp/filament/blob/d51cff6f1f7968aca40f4aea48f0f9dacf03095b/packages/tables/resources/views/columns/text-column.blade.php#L23

So I'm not sure how to avoid this double call.

---
As to how to go from 3 calls to 2 calls, when I look at the compiled blade code it looks like using an expression in a blade parameter calls it twice: once to create the component and once to pass attributes to it? If I turn `:is-click-disabled="$column->isClickDisabled() || $isReordering"` into

```blade
@php
    $isClickDisabled = $column->isClickDisabled() || $isReordering;
@endphp
<x-filament-tables::columns.column
    :column="$column"
    :is-click-disabled="$isClickDisabled"
   ...
/>
```
then I only have two calls. I didn't know this about blade, this could probably be a legitimate PR for a perf improvement to cut computing time by 30% - which I guess is better than nothing.

**Comment by jonaaix** [comment]
@robinmoisson Thanks for your help! I took another look at your findings, and it turns out your original suggestion about CellState was spot on.

One thing I’m still curious about: why isn’t the state initialized as a class member instead of being handled within function calls? While caching would likely have a similar effect, it would be helpful to hear from a maintainer on why it was implemented this way to ensure we’re not unintentionally affecting other functionality.

Coming from an Angular background, one key performance best practice is to avoid calling functions directly from templates (mainly due to the change detection mechanism which would similarly lead in undesired evaluations). Instead, we typically initialize the class with member properties on init and update them granularly on change events, rather than re-evaluating the entire state on each function call. A setup like this might make things more efficient here too.

**Comment by RyanPaiva56** [comment]
I'm running into the same issue where I'm using formatStateUsing to convert pennies to dollars 35000 should be $350 but it runs twice and converts it to $3.50.

```
->formatStateUsing(function (string $state): string {
    logger('formatStateUsing', [$state]);
    return $state ? (string) MoneyHelper::convertToDollars($state) : '';
})
```

Yields:

[2025-03-07 00:09:05] local.DEBUG: formatStateUsing ["35000"] 
[2025-03-07 00:09:05] local.DEBUG: formatStateUsing ["350.00"] 

I just updated composer and I'm running the latest everything on Laravel 11 and Filament 3.

**Comment by danharrin** [comment]
The fix for this issue is being reverted and re-introduced in v4. The issues are all user error, but breaking anyway. See #16165
